### PR TITLE
Feature/simpleusecase aiida

### PR DESCRIPTION
**Pull request by pdiercks**
Hi @dglaeser , Hi @joergfunger ,
this PR adds an implementation of the simple use case with AiiDA. It makes use of the new `shellfunction` functionality (see the README for further info). 

There are still some (minor) problems:

- I had to add optional arguments in `poisson.py` and `postprocessing.py` such that AiiDA would copy the transitive inputs (the .h5 file for example) to the temporary working directory where the process is executed. This also slightly changes the commands that need to be run. 
- AiiDA does not allow for dots in output link labels. This is not really a problem, but something one needs to know. Please see the comments in `simple_use_case.py`
- I could not include the tex compilation step yet. The problem is that `plotoverline.csv` is expected to be in the temporary working directory that will be created to run the process, but currently AiiDA will only transfer that file if that is a direct input, i.e. an argument on the command line.
- the `shellfunction` should allow to easily run shell commands through AiiDA, but it is not designed to be executed on a remote machine. This is explicitly stated in the [AEP](https://github.com/aiidateam/AEP/tree/aep/shell-functions/xxx_shell_functions)

### CODE REVIEW

File: simple_use_case/aiida/simple_use_case.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    # For more information see section "Design choices" > "Output files" in the AiiDA
+    # Enhancement Proposal https://github.com/aiidateam/AEP/tree/aep/shell-functions/xxx_shell_functions
+    arguments = List(
+        ["-2", "-order", "1", "-format", "msh2", "{geometry}", "-o", "mesh"]
+    )
+    results_gmsh = gmsh(arguments=arguments, geometry=geometry)
+
+    # Convert the mesh to XDMF format
+    # The mesh file generated by gmsh is written to a temporary working directory
+    # with a temporary filename. Therefore, meshio can't deduce the input format
```
Comment (joergfunger) [inline simple_use_case/aiida/simple_use_case.py:45]: I would mention that due to the above reasons the output does not have an ending that allows meshio to deduce the type. IMO it is not a temp. file (at least I found that first a little confusing).

File: simple_use_case/aiida/simple_use_case.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    # from the filename and we have to specify this explicitly.
+    arguments = List(["convert", "--input-format", "gmsh", "{mesh}", "mesh.xdmf"])
+    results_meshio = meshio(arguments=arguments, mesh=results_gmsh["mesh"])
+
+    # Solve poisson equation
+    # The additional argument '--h5' is necessary to specify the .h5 file as input,
```
Comment (joergfunger) [inline simple_use_case/aiida/simple_use_case.py:51]: I don't understand this, what is that temp directory?

File: simple_use_case/aiida/simple_use_case.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        script=post_processing_script,
+        pvd=results_poisson["poisson_pvd"],
+        vtu=results_poisson["poisson000000_vtu"],
+    )
+
+    return results_pvbatch
+
+
+if __name__ == "__main__":
+    geometry = SinglefileData(Path("../source/unit_square.geo").resolve())
```
Comment (joergfunger) [inline simple_use_case/aiida/simple_use_case.py:87]: what is that command doing? Could you add some documentation?

File: simple_use_case/aiida/simple_use_case.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    poisson_script = SinglefileData(Path("../source/poisson.py").resolve())
+    post_processing_script = SinglefileData(
+        Path("../source/postprocessing.py").resolve()
+    )
+    results, node = workflow.run_get_node(
```
Comment (joergfunger) [inline simple_use_case/aiida/simple_use_case.py:92]: could you add some documentation here as well (even though that might seem trivial for an expert user)

**Review by joergfunger — APPROVED** [review]
At some point (e.g. at the beginning of the README) it would be good to have a general statement of the other options to implement that (writing a plugin, calcfunction, ..), shortly discuss the pro and con of each approach and highlight, why you choose this approach. Maybe even add links on howto for the other cases.

File: simple_use_case/aiida/simple_use_case.py (hunk 2 above)
Comment (pdiercks) [reply simple_use_case/aiida/simple_use_case.py:51]: The shell command is executed in some temporary directory set as the working directory. See here [code](https://github.com/sphuber/aiida-core/blob/feature/5287/shell-function/aiida/engine/processes/functions/shell.py#L156-L189)

File: simple_use_case/aiida/simple_use_case.py (hunk 1 above)
Comment (pdiercks) [reply simple_use_case/aiida/simple_use_case.py:45]: By now another commit was made, such that the actual filename is used. See this [line of code](https://github.com/sphuber/aiida-core/blob/feature/5287/shell-function/aiida/engine/processes/functions/shell.py#L177)
I will try to make this more intuitive by using `mesh.msh` instead of just `mesh` as the output filename.

File: simple_use_case/aiida/simple_use_case.py (hunk 3 above)
Comment (pdiercks) [reply simple_use_case/aiida/simple_use_case.py:87]: The input (unit_square.geo) has to be declared as a data type known to AiiDA.

File: simple_use_case/aiida/simple_use_case.py (hunk 1 above)
Comment (joergfunger) [reply simple_use_case/aiida/simple_use_case.py:45]: I was just refering to the documention, IMO you don't have to change the code (though any improvement is great)

**Comment by pdiercks** [comment]
Hi @joergfunger , I made the requested changes.

File: simple_use_case/aiida/simple_use_case.py (hunk 2 above)
Comment (joergfunger) [reply simple_use_case/aiida/simple_use_case.py:45]: So in an HPC context the working directory would be the HPC machine that needs to have all the inputs and copy the outputs back. So isn't that then expected? I guess this is somehow confusing with the statement that the shell_function can only be used on a local machine, but maybe the general way of handling inputs/outputs is similar. So I would just write working directory, and the file name IMO is not temporary (the file in the working directory is but not the name).

File: simple_use_case/aiida/simple_use_case.py (hunk 3 above)
Comment (joergfunger) [reply simple_use_case/aiida/simple_use_case.py:83]: I thought this might be difficult to understand for a non-expert Aiida user that looks into our scripts. Could you add the response given here directly as a comment?

File: simple_use_case/aiida/simple_use_case.py (hunk 4 above)
Comment (joergfunger) [reply simple_use_case/aiida/simple_use_case.py:89]: similar to above, it would be nice to shortly add a sentence in the file what run_get_node does.

**Review by joergfunger — APPROVED** [review]

**Comment by dglaeser** [comment]
@pdiercks, thanks for the new implementation. Aiida has no possibility to setup the profile via command-line flags in a non-interactive mode? If so, we have two options to include it in the actions I guess:

- make our own docker image & upload it
- use an existing Aiida image (if that exists)

But the thing with the database is a bummer. I would definitely test, in the actions, if a .pdf file came out of the workflow, and it better be the correct one :D. If we can't get Aiida to work in the actions then that's to me that looks like a sign that its usage is much more difficult than for the others.

But I also find it weird why it is not possible to retrieve the .pdf from the data base as .pdf. Before putting it into the DB it was a valid .pdf file? Whatever that `.handle()` thing is, it seems to be possible for `py2pdf` to convert it into a valid .pdf file - or maybe not, since you say it is empty?

Long story short, I would expect that all our implementations at least work and produce a valid .pdf file, no? We may discuss if the effort to add it to the actions is not worth it, although I would be tempted to say that also that is a hard requirement since otherwise we don't see if everything works (actually, we should probably add a scheduled action run once every week or so)

**Comment by dglaeser** [comment]
@pdiercks, thanks for the new implementation. Aiida has no possibility to setup the profile via command-line flags in a non-interactive mode? If so, we have two options to include it in the actions I guess:

- make our own docker image & upload it
- use an existing Aiida image (if that exists)

But the thing with the database is a bummer. I would definitely test, in the actions, if a .pdf file came out of the workflow, and it better be the correct one :D. If we can't get Aiida to work in the actions then that's to me that looks like a sign that its usage is much more difficult than for the others.

But I also find it weird why it is not possible to retrieve the .pdf from the data base as .pdf. Before putting it into the DB it was a valid .pdf file? Whatever that `.handle()` thing is, it seems to be possible for `py2pdf` to convert it into a valid .pdf file - or maybe not, since you say it is empty?

Long story short, I would expect that all our implementations at least work and produce a valid .pdf file, no? We may discuss if the effort to add it to the actions is not worth it, although I would be tempted to say that also that is a hard requirement since otherwise we don't see if everything works (actually, we should probably add a scheduled action run once every week or so)

File: simple_use_case/aiida/suc_v2b1.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+with paper["paper_pdf"].open(mode="rb") as handle:
+    output_file.write(handle.read())
+
+pdfWriter.write(output_file)
+
+breakpoint()
+# while at breakpoint, try viewing `target.pdf` --> not empty
+# after c(ontinue) and end of program `target.pdf` is empty ...
+output_file.close()
```
Comment (pdiercks) [inline simple_use_case/aiida/suc_v2b1.py:133]: Hi @dglaeser 
to clarify: the produced output PDF is the correct one and while the program is interrupted I can view "target.pdf" and it is not empty. However, after the program exits and I view "target.pdf" again it is empty ... (not sure what is happening)

Comment (pdiercks) [reply simple_use_case/aiida/suc_v2b1.py]: I fixed it. I should have just done that before posting, sorry.

**Comment by pdiercks** [comment]
Regarding the github actions:

1. I will try to install everything as described in the README and see if I can configure everything non-interactively.
2. If that does not work: there is an existing docker image that we could use. It also has conda installed, but installing all deps for the simple use case (ie fenics etc) on top of aiida did not work for me. Conda is not able to resolve the conflicts and this is a problem I am not sure it's worth digging into. I would rather install the simple use case deps in an separated environment and use `conda run -n <env> <any-command>`.

File: simple_use_case/aiida/suc_v2b1.py (hunk 5 above)
Comment (dglaeser) [reply simple_use_case/aiida/suc_v2b1.py]: great, thanks a lot!!

File: .github/actions/install-aiida-services/action.yml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+name: install-aiida-services
```
Comment (dglaeser) [inline .github/actions/install-aiida-services/action.yml:1]: it seems like these actions are not used in the main .yml file or am I mistaking?

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -10,131 +10,114 @@ on:
...
+    - name: install-aiida-shell
+      working-directory: simple_use_case/aiida
+      shell: bash -l {0}
+      run: |
+        git clone https://github.com/sphuber/aiida-shell.git
+        cd aiida-shell
+        git checkout v0.1.0
+        pip install -e .
+
+    - name: check-conda-env
```
Comment (dglaeser) [inline .github/workflows/main.yml:84]: is this job and the `check-status` job left over from debugging or do these checks prevent something? I would expect that if these two jobs failed, then the pipeline would also fail without these two jobs existing? If so, could we delete them (the action is already quite long & complex)?

**Comment by dglaeser** [comment]
Thanks a lot @pdiercks !! I just have two minor points (see comments).

File: .github/actions/install-aiida-services/action.yml (hunk 6 above)
Comment (pdiercks) [reply .github/actions/install-aiida-services/action.yml]: yes, you are right. I removed them.

File: .github/workflows/main.yml (hunk 7 above)
Comment (pdiercks) [reply .github/workflows/main.yml]: Yes, these jobs were left over from debugging.

**Comment by pdiercks** [comment]
I feel like this is ready to merge. Are there further changes required?

**Comment by dglaeser** [comment]
>  I feel like this is ready to merge. Are there further changes required?
Not from my side, thanks for addressing the points I raised!
